Add sqmr

ikligge opened the pull request:
Dear developers,

this is finally my rewritten SQMR iteration (you may remember my post in deal.II user group long ago). I mainly changed the iteration itself, the AdditionalData and some variables, but almost nothing of the remaining/surrounding code. The benefits of my changes are that the new algorithm will perform much faster and is adaptable to left and right side preconditioning of the system matrix. I also tried to put as many comments as necessary on the actual code, but haven't Doxygen-tested them, so there may be some corrections necessary. In addition I didn't have the opportunity to test the solver on diverse problems, but it worked out quite fine on my saddle point system (just symmetric) and the local blocks (symmetric and positive definite), so I guess the algorithm will be correct somehow.

Furthermore I would like to suggest a renaming of the solver. The original creator of the QMR and every related method, Freund/Nachtigal et al., also proposed a "QMR Squared", which is usually abbreviated by QMRS. I think, SQMR is a more common name, too.

Yours sincerely,
Ingo Kligge

davydden commented:
please run the indentation script `./contrib/utilities/indent` with `astyle@2.04` in path.

masterleinad commented:
./run-tests

bangerth commented:
@ikligge -- first, many thanks for implementing this and proposing it as a patch! One can never have enough solver options! I also particularly appreciate the effort you've put into documenting this class better -- well done!

I've got a couple of comments worth discussing before we move on to a detailed review:

* @davydden already mentioned the indentation script. It looks like you already figured out how this works, right?
* You did the development based on an old version of the file, and the pull request undoes some development that has happened in this file since you started your work. I would like to find a way to make sure that the things that were changed in master get put back in with your patch. Do you know how to do this? On way to see what has happened to a file is using the `gitk` GUI, for example by calling it with `gitk include/deal.II/lac/solver_qmrs.h`. If you remember when you started your work, you can see with the GUI what patches have happened since then, and maybe find a way to manually re-apply them -- unfortunately, they look like they conflict with your work, so some manual work is necessary. If you want more help, let us know and we'll be happy to assist!

In any case, thanks already! Your contribution is much appreciated!

ikligge commented:
@davydden @bangerth Yes, I somehow established to run the indentation script. This should be no matter anymore in future commits. 

@masterleinad I think, it may be wise to postpone any test right now...

@bangerth Yes, I see some issues. They seem to be mostly minor things which (hopefully) can easily be corrected. I will have a closer look tomorrow. Some tests afterwards will be wise then, I think...

bangerth commented:
Great, thanks for being willing to work through these merge issues. I agree that we should first get the code reviewed, and then we can add tests to the pull request.

bangerth commented:
@ikligge -- any news? Let us know if you need help with anything!

ikligge commented:
@bangerth Sorry about the delay. Yes, several problems. Compiling the deal.II libraries ("make install") freezes my laptop (very old machine) or at least my Ubuntu graphical surface, so I need to fall back on university servers, installing and testing my changes there. I do so right now and I hope it'll finally work...

ikligge commented:
Okay, this is the current version with all changes applied (and with indent-script run). But it isn't really tested. I tried to run tutorial step 5 but it didn't finish. It could be the usual unstableness of the method itself - don't know, haven't seen the log history and haven't spent much time on it. I can try to make it run with my master thesis code, but this will be again quite nasty work of back and forth transferring between my laptop and the university system and I cannot guarantee that all changes will be applicable. Has anyone of you some nice example ready for some easy plug-in solving in store?

bangerth commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -74,86 +94,101 @@ class SolverQMRS : public Solver<VectorType>
-    {}
+    AdditionalData (bool left_preconditioning = false,
+                    double solver_tolerance = 1.e-9,
+                    bool breakdown_testing = true,
```
This is an incompatible change from the old list of arguments to the new one. That's ok, but it should be mentioned in a changelog entry in `doc/news/changes/incompatible`.

Also, since all of these arguments are input arguments, can you mark them as `const`?

ikligge replied:
So small changelog textfile written and every input argument `const` from now on.

bangerth commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -205,67 +239,56 @@ class SolverQMRS : public Solver<VectorType>
                                const PreconditionerType &preconditioner)
 {
-  LogStream::Prefix prefix("QMRS");
+  LogStream::Prefix prefix("SQMR"); //TODO Renaming issue
```
The change is ok, but please remove the `TODO` marker

ikligge replied:
Okay, done.

bangerth commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -274,129 +297,167 @@ SolverQMRS<VectorType>::solve (const MatrixType         &A,
 
-  p = v;
+  tau = t.norm_sqr ();
+  res = sqrt (tau);
```
Please use `std::sqrt`

ikligge replied:
Okay, done.

bangerth commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -274,129 +297,167 @@ SolverQMRS<VectorType>::solve (const MatrixType         &A,
+      // Right preconditioning
+      t = r;
+      preconditioner.vmult (q, t);
+    };
```
no need for the semicolon

ikligge replied:
Oups, yes. Done

ikligge commented:
I don't have time today, but I think I'm going to try to run some tests tomorrow, especially with my master thesis code, for which I already know the SQMR behaviour.

bangerth commented:
OK, thanks. Let us know what you find in your tests!

Since you say that you've made all of the requested modifications, can you push these to github so they show up here as well?

ikligge commented:
Okay, here it is. It worked with my master thesis code, so it should be working with other suitable problems as well I suppose. I haven't run any official deal.II test yet because (I think) my local clone is interfering with the setup of the older institute version which is set on a (institute-)central fileserver.

bangerth commented:
Let's see what the regression tester uncovers!

/run-tests

bangerth commented:
There are 4 tests that fail:
http://www.math.clemson.edu/~heister/cib/dealii-tjhei-gcc-serial/3fedae767307beda9cffbb1613d6ed7ecbb4cd82/log-gcc

I think that most of these differences are probably harmless, but the output shown on the page may not show enough data. Do you know how to update test outputs? If not, let me know and I can take things from here!

ikligge commented:
No sorry, I don't know. I have to admit, that I don't really know anything about the test suites.

bangerth commented:
@ikligge: I ran the testsuite and there is one new failing test, `lac/linear_operator_10a.debug`, that yields different output:
```
140,142c140,142
< DEAL:Solvers:SolverQMRS:S_Op::0.000000000 0.5000000000 1.000000000 1.500000000 2.000000000 2.500000000 3.000000000 3.500000000 4.000000000 4.500000000 
< DEAL:Solvers:SolverQMRS:C_Op::0.000000000 0.5000000000 1.000000000 1.500000000 2.000000000 2.500000000 3.000000000 3.500000000 4.000000000 4.500000000 
< DEAL:Solvers:SolverQMRS:C_Op2::0.000000000 0.5000000000 1.000000000 1.500000000 2.000000000 2.500000000 3.000000000 3.500000000 4.000000000 4.500000000 
---
> DEAL:Solvers:SolverQMRS:S_Op::0.000000000 1.000000000 2.000000000 3.000000000 4.000000000 5.000000000 6.000000000 7.000000000 8.000000000 9.000000000 
> DEAL:Solvers:SolverQMRS:C_Op::0.000000000 2.000000000 4.000000000 6.000000000 8.000000000 10.00000000 12.00000000 14.00000000 16.00000000 18.00000000 
> DEAL:Solvers:SolverQMRS:C_Op2::0.000000000 2.000000000 4.000000000 6.000000000 8.000000000 10.00000000 12.00000000 14.00000000 16.00000000 18.00000000 
```
Do you happen to know why this is? Can you see what might cause the difference? (Top: new output. Bottom: old output.)

ikligge commented:
I don't know why this happened. But is this difference between old and new necessarily a bug? I've scrolled through my versions of `lac/linear_operator_10a.cc` and the output file `linear_operator_10a.with_trilinos=on.with_64bit_indices=off.output` and the new output looks like the correct solution of the test (which any other solver "scored" by the way).

bangerth commented:
Yes, this is correct -- the old output could also have been wrong. So do I read your message correctly that (i) you believe that the new output is correct, (ii) that the new output is also what is produced by other solvers, and (iii) that the old output was consequently wrong?

If that is the case, would you mind updating the stored output file for the test and add this to the pull request?

It would of course be interesting to know what part of the patch made the output correct. If I recall correctly, your patch is really only intended to extend the solver, not fix anything. But I recognize that it may not be easy to figure out what exactly has changed, and it may not be important either if you are convinced that the new output is correct anyway.

jppelteret commented:
> Yes, this is correct -- the old output could also have been wrong. 

I've noted that this is a test that I wrote. I see that the original solution output by the solver was indeed incorrect. The solution given by CG is
```
DEAL:Solvers:SolverCG:S_Op::0.000000000 0.5000000000 1.000000000 1.500000000 2.000000000 2.500000000 3.000000000 3.500000000 4.000000000 4.500000000 
DEAL:Solvers:SolverCG:C_Op::0.000000000 0.5000000000 1.000000000 1.500000000 2.000000000 2.500000000 3.000000000 3.500000000 4.000000000 4.500000000 
DEAL:Solvers:SolverCG:C_Op2::0.000000000 0.5000000000 1.000000000 1.500000000 2.000000000 2.500000000 3.000000000 3.500000000 4.000000000 4.500000000 
```
which means that the new output matches those of the other solvers (confirming what @ikligge said). Nice work on the bug fix!

jppelteret commented on tests/lac/linear_operator_08.with_umfpack=off.with_lapack=true.output:
```diff
@@ -47,24 +47,24 @@ DEAL:Solvers:Linear operator:cg::Starting value 16.88194302
-DEAL:Solvers:Standard solver:QMRS::Convergence step 1 value 0.000000000
-DEAL:Solvers:Linear operator:QMRS::Starting value 16.88194302
-DEAL:Solvers:Linear operator:QMRS::Convergence step 1 value 0.000000000
+DEAL:Solvers:Standard solver:SQMR::Starting value 16.88194302
```
I thought I had a dyslexic moment when writing this test, but I note that the solver is indeed called `SolverQMRS`. So what's with this change (and the similar changes to the other tests)?

ikligge replied:
Oh, this is just a naming issue. The original creators of the algorithm (Roland Freund and Noel Nachtigal) used the short form SQMR for this method ("QMR for symmetric matrices") and QMRS for another, different method ("QMR squared", named similarly to CGS). I think it's better to use the conventional name to avoid possible confusion. I at least changed the deallog name, thus these differences.

masterleinad commented:
From my point of view, this should be ready for merging.
@ikligge Do you know how to squash commits (interactive rebase)? If so, would you mind to reduce the number of (correctly compiling) commits?

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -31,15 +31,40 @@ DEAL_II_NAMESPACE_OPEN
+ * indefinite linear systems, NASA STI/Recon Technical Report N, 95 (1994) and
+ * Freund/Nachtigal: Software for simplified Lanczos and QMR algorithms,
+ * Appl. Num. Math. 19 (1995), pp. 319-341 and provides both right and left
+ * but not split preconditioning.
```
but not split -> (but not split)

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -31,15 +31,40 @@ DEAL_II_NAMESPACE_OPEN
+ *
+ * <h3>Trade off of stability to simplicity</h3>
+ *
+ * Note, that the QMR implementation the given algorithm is based on, is
```
Note that the QMR implementation that the given algorithm is based on is

jppelteret replied:
Also, the abbreviation "QMR" has not been previously defined

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -31,15 +31,40 @@ DEAL_II_NAMESPACE_OPEN
+ * derived from classical BiCG. It can be shown (Freund/Szeto: A transpose-free
+ * quasi-minimal residual squared algorithm for non-Hermitian linear systems,
+ * Advances in Computer Methods for Partial Differential Equations VII
+ * (IMACS, New Brunswick, NJ, 1992) pp. 258-264), that the QMR iterates can
```
pp. 258-264) that the

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -31,15 +31,40 @@ DEAL_II_NAMESPACE_OPEN
+ * some scalar updates. Possible breakdowns (or precisely, divisions by
+ * zero) of BiCG therefore obviously transfer to this simple no-look-ahead algorithm.
+ *
+ * In return the algorithm is cheap compared to classical QMR or BiCGStab
```
or BiCGStab,

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -31,15 +31,40 @@ DEAL_II_NAMESPACE_OPEN
+ * zero) of BiCG therefore obviously transfer to this simple no-look-ahead algorithm.
+ *
+ * In return the algorithm is cheap compared to classical QMR or BiCGStab
+ * using only one matrix-vector-product with the system matrix and
```
matrix-vector product

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -31,15 +31,40 @@ DEAL_II_NAMESPACE_OPEN
+ * one application of the preconditioner per iteration respectively.
+ *
+ * The residual used for measuring convergence is only approximately calculated
+ * by an upper bound. If this value comes below a threshold given by the
```
given by -> prescribed within

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -31,15 +31,40 @@ DEAL_II_NAMESPACE_OPEN
+ * AdditionalData struct, then the exact residual of the current
+ * QMR iterate will be calculated using another multiplication with the system
+ * matrix. By experience (according to Freund and Nachtigal) this technique
+ * is useful for a threshold that is ten times the solving tolerance and in
```
tolerance, and in

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -65,7 +85,7 @@ DEAL_II_NAMESPACE_OPEN
  *
  *
- * @author Guido Kanschat, 1999
+ * @author Guido Kanschat, Ingo Kligge 1999, 2017
```
May you please write the author's list as 

> Guido Kanschat, 1999;  Ingo Kligge 2017

Although we're not completely consistent throughout the library, tend to extend the author's list in this way.

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -74,86 +94,101 @@ class SolverQMRS : public Solver<VectorType>
-   * another matrix vector multiplication.
+   * The user is able to switch between right and left preconditioning, that means
+   * solving the systems <i>P<sup>-1</sup>A</i> and <i>AP<sup>-1</sup></i> respectively,
+   * using the corresponding parameter. Note, that left preconditioning means to
```
Note that left

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -74,86 +94,101 @@ class SolverQMRS : public Solver<VectorType>
+   * The user is able to switch between right and left preconditioning, that means
+   * solving the systems <i>P<sup>-1</sup>A</i> and <i>AP<sup>-1</sup></i> respectively,
+   * using the corresponding parameter. Note, that left preconditioning means to
+   * employ the preconditioned (BiCG-)residual and otherwise the unpreconditioned one.
```
I think that "otherwise the unpreconditioned one" is a bit ambiguous. Do you mean

> (BiCG-)residual, and right preconditioning implies the use of the unpreconditioned (BiCG-)residual.

?

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -74,86 +94,101 @@ class SolverQMRS : public Solver<VectorType>
+   * The default is the application from the right side.
    *
-   * QMRS, is susceptible to breakdowns, so we need a parameter telling us,
+   * The solver_tolerance threshold is used to define the said bound below which the residual
```
@p solver_tolerance

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -74,86 +94,101 @@ class SolverQMRS : public Solver<VectorType>
    *
-   * QMRS, is susceptible to breakdowns, so we need a parameter telling us,
+   * The solver_tolerance threshold is used to define the said bound below which the residual
+   * is computed exactly. See class documentation for more information. The default value is 1e-9,
```
the class documentation

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -74,86 +94,101 @@ class SolverQMRS : public Solver<VectorType>
+   * is computed exactly. See class documentation for more information. The default value is 1e-9,
+   * that is the default solving precision multiplied by ten.
+   *
+   * SQMR is susceptible to breakdowns (divisions by zero), so we need a parameter telling us,
```
so we need a parameter telling us which

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -74,86 +94,101 @@ class SolverQMRS : public Solver<VectorType>
    * which numbers are considered zero. The proper breakdown criterion is very
-   * unclear, so experiments may be necessary here.
+   * unclear, so experiments may be necessary here. It is even possible to achieve convergence
+   * despite of dividing through small numbers. There are even cases, in which it is advantageous to
```
dividing through by small numbers

jppelteret replied:
even cases in which

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -74,86 +94,101 @@ class SolverQMRS : public Solver<VectorType>
+   * unclear, so experiments may be necessary here. It is even possible to achieve convergence
+   * despite of dividing through small numbers. There are even cases, in which it is advantageous to
+   * accept such divisions because the cheap iteration cost makes the algorithm the fastest of all
+   * available indefinit solvers. Nonetheless, the default breakdown threshold value is 1e-16.
```
indefinit -> indefinite (or better, if correct, "iterative")

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -74,86 +94,101 @@ class SolverQMRS : public Solver<VectorType>
 
     /**
-     * Breakdown threshold.
+     * Representing the threshold below which the current residual is computed exactly.
```
The threshold below which

jppelteret commented on doc/news/changes/incompatibilities/20171206IngoKligge:
```diff
@@ -0,0 +1,5 @@
+Changed: The parameters contained in struct SolverQMRS::AdditionalData has
```
have been changed

jppelteret commented on doc/news/changes/incompatibilities/20171206IngoKligge:
```diff
@@ -0,0 +1,5 @@
+Changed: The parameters contained in struct SolverQMRS::AdditionalData has
+been changed in order to meet the rewritten algorithm. See class documentation 
```
See the class documentation

jppelteret commented on include/deal.II/lac/solver_qmrs.h:
```diff
@@ -74,86 +94,101 @@ class SolverQMRS : public Solver<VectorType>
      * Constructor.
      *
-     * The default is no exact residual computation and breakdown parameter
+     * The default is right preconditioning and breakdown parameter
```
The default is the use right preconditioning, with the @p solver_tolerance chosen to be 1e-9 and the @p breakdown_threshold set at 1e-16.

jppelteret commented:
@ikligge Great, thanks a lot! Yes, it would be appropriate to add an entry in `doc/news/changes/minor` stating that the algorithm has been refactored and made more robust.

If you want to squash this all into one commit then that would be fine. Otherwise you could make the changes to the main class (plus changelogs) one commit and the updates to the test outputs a second commit.

masterleinad commented:
> About the changelog entry: do you mean a entry which states that the algorithm itself has changed?

That should be something similar to:
> Improved: SolverQMRS was renamed to SolverSQMR. Furthermore, the new implementation performs much faster and is adaptable to left and right side preconditioning of the system matrix. [...]

> If you want to squash this all into one commit then that would be fine. Otherwise you could make the changes to the main class (plus changelogs) one commit and the updates to the test outputs a second commit.

I agree with both of these suggestions. 
If you have never done such an interactive rebase before, I would suggest that you save the current branch into a backup branch first. Then you are on the safe side if something goes wrong.

ikligge commented:
Okay, I've added another changelog file and squashed some commits together. But how to push this to GitHub? My local changes are rejected because Git thinks my local branch has fallen behind to the online one. Shall I use force (`git push --force ...`)?

jppelteret commented:
@ikligge Thanks for pushing your changes again. Unfortunately your branch is based off quite an old version of the `master` branch, so there are some conflicts that need to be resolved before we can merge this. Do you know how to rebase off of the latest version of `master`? If not, then I can sketch out how you'd do this. 

Also, I noticed that your commit 65226ab contains a lot of information in the message, some of which I don't think is relevant anymore (I'm guessing you followed my literal instruction of using [squash rather than fixup](https://stackoverflow.com/a/16758662), which would have merged all of the commit messages into one). Would it be possible to tidy that up a little as well? When doing an interactive rebase its possible to edit the commit messages. Again, if you need a hand with this then just let me know.

Thanks for working with me through this - we're nearly there :-)

jppelteret commented:
@ikligge I think that you also forgot to create an additional changelog entry in `doc/news/changes/minor`. Don't be modest about your contribution :-)

ikligge commented:
@jppelteret I cleaned up the commit message by another rebase with `reword` command. Should be upcoming with the force push I will have to do after the next rebase on top of the `master` branch. Speaking of which, I found [this rebase-Tutorial](https://robots.thoughtbot.com/git-interactive-rebase-squash-amend-rewriting-history) (almost of the bottom of the page, above the danger paragraph). Shall I do it this way? But what is `upstream` in this case here?

Isn't there a `doc/news/changes/minor` log entry? I've actually added one...

jppelteret commented:
@ikligge Great, that part of the tutorial you linked to (specifically "Rebase on top of master ") describes exactly how it needs to be done. In this case `upstream` is `https://github.com/dealii/dealii.git` as you're wanting to synchronise with the current state of our remote repository. I've done a dry-run of this process on my side and there shouldn't be any conflicts to fix, so hopefully that makes things easier for you.

Maybe you created the extra changelog entry but forgot to commit it. I don't see it in the [current list of changed files](https://github.com/dealii/dealii/pull/5536/files).

ikligge commented:
Okay, I hope everything will be in place now.

jppelteret commented:
>Okay, I hope everything will be in place now.

Nearly :-) Can you check to see if you've committed and pushed the changelog entry? Its still not being reflected in the PR.

jppelteret commented:
Awesome. Thanks for working through this with me. I appreciate your patience!

ikligge commented:
Thank you for your help, @jppelteret ! And also thanks to the others (if they'll read this...)!

jppelteret commented:
Thanks a lot for your contribution @ikligge! I appreciate you sticking it out with me, and I hope it won't be the last addition that you make to deal.II!

ikligge commented:
It's my pleasure. Yes, if I find a point to step in again, I will be glad to offer another contribution.

bangerth commented:
Thanks also from my side for your patience. Things don't usually take this long to get merged, and I hope we didn't scare you off! It gets easier with time as you learn the various processes and styles of the project. We're definitely looking forward to more contributions!
